### PR TITLE
Support multi-GPU training for supervised training

### PR DESCRIPTION
**Pull request by ksikka**

### CODE REVIEW

File: docs/roadmap.md [hunk 1]
```diff
@@ -1,7 +1,8 @@
 # Lightning Pose development roadmap
 
 ## General enhancements
-- [ ] multi-GPU training for supervised models
+- [x] multi-GPU training for supervised models
```
Comment (themattinthehatt) [inline docs/roadmap.md:4]: add link to PR (for example see line 9)

File: docs/source/faqs.rst [hunk 2]
```diff
@@ -65,12 +65,21 @@ FAQs
...
+    Temporal Context Network model, multi-view datasets, or high-resolution images. For this reason
+    we recommend using a GPU with a minimum of 8GB of memory, but preferrably 16GB.
+    
+    Some users using a combination of the memory-intensive features above may still run into issues. 
+    There are a few techniques available to reduce the memory consumption:
+    
+    * Reduce ``train_batch_size``. Memory usage is directly proportional to batch size.
+    * Enable multi-GPU training (only supported for supervised training) using ``num_gpus``.
+    * Reduce image resolution using ``image_resize_dims``.
+    * Enable gradient accumulation using ``accumulate_grad_batches``.
```
Comment (themattinthehatt) [inline docs/source/faqs.rst:78]: add that the `accumulate_grad_batches` field will need to be added to the `training` section of the config

File: lightning_pose/models/base.py [hunk 3]
```diff
@@ -457,12 +457,16 @@ def evaluate_labeled(
 
         if stage:
             # log overall supervised loss
-            self.log(f"{stage}_supervised_loss", loss, prog_bar=True)
+            self.log(f"{stage}_supervised_loss", loss, prog_bar=True, sync_dist=True)
```
Comment (themattinthehatt) [inline lightning_pose/models/base.py:460]: add comment about what `sync_dist=True` does

File: lightning_pose/utils/scripts.py [hunk 4]
```diff
@@ -135,14 +137,37 @@ def get_data_module(
...
+        # size in a multi-gpu setting.
+        if cfg.training.train_batch_size % cfg.training.num_gpus != 0:
+            raise ValidationError(
+                "train_batch_size should be a multiple of num_gpus. "
+                "train_batch_size=%d, num_gpus=%d"
+                % (cfg.training.train_batch_size, cfg.training.num_gpus)
+            )
+        if cfg.training.val_batch_size % cfg.training.num_gpus != 0:
+            raise ValidationError(
+                "val_batch size should be a multiple of num_gpus. "
```
Comment (themattinthehatt) [inline lightning_pose/utils/scripts.py:160]: small stylistic thing, I'd like to eventually update the codebase to only use f-strings, they're cleaner in my opinion, e.g.
```
f"val_batch_size={cfg.training.val_batch_size}, num_gpus={cfg.training.num_gpus}"
```

File: tests/utils/test_scripts.py [hunk 5]
```diff
@@ -158,3 +159,72 @@ def test_export_predictions_and_labeled_video(
 
     # clean up logging
     remove_logs()
+
+
+def _supervised_multi_gpu_cfg(cfg):
```
Comment (themattinthehatt) [inline tests/utils/test_scripts.py:164]: this is fine, but you could also make a variable `cfg_multigpu` as a fixture in the file `tests/conf.py` and pass it to whatever test(s) you need

File: tests/utils/test_scripts.py [hunk 6]
```diff
@@ -158,3 +159,72 @@ def test_export_predictions_and_labeled_video(
...
+def test_get_data_module_num_gpus_0(cfg):
+    cfg = _supervised_multi_gpu_cfg(cfg)
+    # when num_gpus is set to 0, i.e. from a deprecated config
+    cfg.training.num_gpus = 0
+    data_module = get_data_module(cfg, Mock(spec=BaseTrackingDataset))
+
+    # assert num_gpus gets modified to 1
+    assert cfg.training.num_gpus == 1
+    # the rest of the behavior follows correctly
+    assert data_module.train_batch_size == 4
```
Comment (themattinthehatt) [inline tests/utils/test_scripts.py:192]: compare to value in `cfg` rather than hard code (same for two below)

File: tests/utils/test_scripts.py [hunk 7]
```diff
@@ -158,3 +159,72 @@ def test_export_predictions_and_labeled_video(
...
+    assert data_module.val_batch_size == 16
+    assert data_module.test_batch_size == 16
+
+
+def test_get_data_module_multi_gpu_batch_size_adjustment(cfg):
+    cfg = _supervised_multi_gpu_cfg(cfg)
+    data_module = get_data_module(cfg, Mock(spec=BaseTrackingDataset))
+    # train, val batch sizes should be divided by num_gpus
+    assert data_module.train_batch_size == 2
```
Comment (themattinthehatt) [inline tests/utils/test_scripts.py:201]: same comment as above, pull `train_batch_size` and `num_gpus` out of cfg

**Review by themattinthehatt — CHANGES_REQUESTED** [review]
looks great, most requests are about comments/docs

File: lightning_pose/models/base.py (hunk 3 above)
Comment (ksikka) [reply lightning_pose/models/base.py:463]: Added

File: tests/utils/test_scripts.py (hunk 5 above)
Comment (ksikka) [reply tests/utils/test_scripts.py:164]: Acknowledged. I'm not sure I'll use this is any other test yet.

File: lightning_pose/utils/scripts.py [hunk 8]
```diff
@@ -152,11 +177,14 @@ def get_data_module(
         )
     else:
         if cfg.model.model_type == "heatmap_mhcrnn" and cfg.dali.context.train.batch_size < 5:
-            raise ValueError(
+            raise ValidationError(
```
Comment (ksikka) [inline lightning_pose/utils/scripts.py:180]: @themattinthehatt I changed your ValueError to ValidationError for consistency. My only reason for preferring validation error is it looks more like a user error to the end user.

File: docs/source/faqs.rst [hunk 9]
```diff
@@ -65,12 +65,22 @@ FAQs
...
+    we recommend using a GPU with a minimum of 8GB of memory, but preferrably 16GB.
+    
+    Some users using a combination of the memory-intensive features above may still run into issues. 
+    There are a few techniques available to reduce the memory consumption:
+    
+    * Reduce ``train_batch_size``. Memory usage is directly proportional to batch size.
+    * Enable multi-GPU training (only supported for supervised training) using ``num_gpus``.
+    * Reduce image resolution using ``image_resize_dims``.
+    * Enable gradient accumulation using ``accumulate_grad_batches``. This parameter is not included
+      in the config by default and should be added manually to the ``training`` section.
```
Comment (ksikka) [inline docs/source/faqs.rst:79]: @themattinthehatt  Added "this parameter is not included..." just now.

File: docs/source/user_guide/config_file.rst [hunk 10]
```diff
@@ -70,6 +70,15 @@ Below is a list of some commonly modified arguments related to model architectur
   Therefore, 300 epochs, at 40 batches per epoch, is equal to 300*40=12k total batches
   (or iterations).
 
+* ``training.num_gpus``: enables multi-GPU training, only supported for supervised losses
+  (``losses_to_use: []``) . ``train_batch_size`` and ``val_batch_size`` must be divisible by
+  ``num_gpus`` as each batch will be divided amongst the GPUs.
+
+* ``training.accumulate_grad_batches``: (experimental) number of batches to accumulate gradients
+  for before updating weights. Simulates larger batch sizes with memory-constrained GPUs. This
+  parameter is not included in the config by default and should be added manually to the
```
Comment (ksikka) [inline docs/source/user_guide/config_file.rst:79]: @themattinthehatt  Added "this parameter is not included..." just now.

File: docs/source/faqs.rst (hunk 2 above)
Comment (ksikka) [reply docs/source/faqs.rst]: Done

File: lightning_pose/utils/scripts.py (hunk 4 above)
Comment (ksikka) [reply lightning_pose/utils/scripts.py]: Ah, didn't know about that. Cool feature. Updated to use that.

File: tests/utils/test_scripts.py (hunk 7 above)
Comment (ksikka) [reply tests/utils/test_scripts.py]: Done

File: tests/utils/test_scripts.py (hunk 6 above)
Comment (ksikka) [reply tests/utils/test_scripts.py]: Done

**Review by themattinthehatt — APPROVED** [review]

**Pull request merged**